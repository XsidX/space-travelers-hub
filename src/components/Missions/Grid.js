import React from 'react';
import propTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { missionsActions } from '../../store/missions/missions-slice';

const GridHead = ({ children }) => (
  <div className="grid">
    <div className="grid__item grid__mission">
      <h3 className="grid__item--title">Mission</h3>
    </div>
    <div className="grid__item grid__description">
      <h3 className="grid__item--title">Description</h3>
    </div>
    <div className="grid__item grid__status">
      <h3 className="grid__item--title">Status</h3>
    </div>
    {children}
  </div>
);

const GridRow = ({ mission }) => {
  const dispatch = useDispatch();
  const { id, name, description, joined } = mission;

  const toggleJoin = () => {
    dispatch(missionsActions.missionJoined(id));
  };

  return (
    <>
      <div className="grid__item grid__mission ">
        <h3 className="grid__item--title">{name}</h3>
      </div>
      <div className="grid__item grid__description">
        <h3>{description}</h3>
      </div>
      <div className="grid__item grid__status">
        <button
          className={`${joined && 'status-active'} status-btn`}
          type="button"
        >
          {joined ? 'Active Member' : 'NOT A MEMBER'}
        </button>
        <button
          className={`${joined && 'outline-red'} action-btn-outline`}
          type="button"
          onClick={toggleJoin}
        >
          {joined ? 'Leave Mission' : 'Join Mission'}
        </button>
      </div>
    </>
  );
};

GridHead.propTypes = {
  children: propTypes.node.isRequired,
};

GridRow.propTypes = {
  mission: propTypes.shape({
    id: propTypes.string.isRequired,
    name: propTypes.string.isRequired,
    description: propTypes.string.isRequired,
    joined: propTypes.bool,
  }).isRequired,
};

export { GridHead, GridRow };
