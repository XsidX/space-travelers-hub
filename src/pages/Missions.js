import React from 'react';

const Missions = () => (
  <div>
    <h1>Missions</h1>
  </div>
);

export default Missions;
