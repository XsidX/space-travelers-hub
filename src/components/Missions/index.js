import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { GridHead, GridRow } from './Grid';
import { getMissions } from '../../store/actions';

const Missions = () => {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions) || [];

  useEffect(() => {
    dispatch(getMissions());
  }, []);

  return (
    <GridHead>
      {missions.map((mission) => (
        <GridRow key={mission.id} mission={mission} />
      ))}
    </GridHead>
  );
};

export default Missions;
