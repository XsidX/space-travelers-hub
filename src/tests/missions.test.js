import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import * as reactRedux from 'react-redux';
import { GridRow } from '../components/Missions/Grid';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

describe('GridRow component', () => {
  const mockState = [
    {
      id: '1',
      name: 'Test Mission',
      description: 'Test description',
      joined: false,
    },
  ];

  const dispatch = reactRedux.useDispatch;

  beforeEach(() => {
    dispatch.mockImplementation(() => jest.fn());
  });

  afterEach(() => {
    dispatch.mockClear();
  });

  it('renders not a member if joined property is false', () => {
    render(
      mockState.map((mission) => (
        <GridRow key={mission.id} mission={{ ...mission, joined: false }} />
      )),
    );

    const activeMemberButton = screen.getAllByText('NOT A MEMBER', {
      exact: false,
    });

    expect(activeMemberButton[1]).toBeInTheDocument();
  });

  it('renders active member if joined property is true', () => {
    render(
      mockState.map((mission) => (
        <GridRow key={mission.id} mission={{ ...mission, joined: true }} />
      )),
    );

    const activeMemberButton = screen.getAllByText('Active Member', {
      exact: false,
    });

    expect(activeMemberButton[1]).toBeInTheDocument();
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        mockState.map((mission) => (
          <GridRow key={mission.id} mission={{ ...mission, joined: false }} />
        )),
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
