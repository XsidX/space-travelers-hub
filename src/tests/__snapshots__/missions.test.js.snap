// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridRow component renders correctly 1`] = `
Array [
  <div
    className="grid__item grid__mission "
  >
    <h3
      className="grid__item--title"
    >
      Test Mission
    </h3>
  </div>,
  <div
    className="grid__item grid__description"
  >
    <h3>
      <button
        className="false status-btn grid__description--member-status"
        type="button"
      >
        NOT A MEMBER
      </button>
      Test description
    </h3>
  </div>,
  <div
    className="grid__item grid__status"
  >
    <button
      className="false status-btn grid__item--member-status"
      type="button"
    >
      NOT A MEMBER
    </button>
    <button
      className="false action-btn-outline"
      onClick={[Function]}
      type="button"
    >
      Join Mission
    </button>
  </div>,
]
`;
